test/opt/fold_test.cpp

```diff
@@ -1016,10 +1034,238 @@ INSTANTIATE_TEST_SUITE_P(TestCase, IntegerInstructionFoldingTest,
+          "%2 = OpUConvert %uint %short_n5\n" +
+          "OpReturn\n" +
+          "OpFunctionEnd",
+      2, 65531),
+  // Test case 74: Fold short(-24(int)). The upper bits should be cleared. So 0xFFFFFFE8 should become 0x0000FFE8.
+  InstructionFoldingCase<uint32_t>(
+      Header() + "%main = OpFunction %void None %void_func\n" +
+          "%main_lab = OpLabel\n" +
+          "%2 = OpUConvert %ushort %int_n24\n" +
+          "OpReturn\n" +
+          "OpFunctionEnd",
+      2, 65512)
 ));
 // clang-format on
 
+using LongIntegerInstructionFoldingTest =
+    ::testing::TestWithParam<InstructionFoldingCase<uint64_t>>;
+
+TEST_P(LongIntegerInstructionFoldingTest, Case) {
+  const auto& tc = GetParam();
+
+  std::unique_ptr<IRContext> context;
+  Instruction* inst;
+  std::tie(context, inst) =
+      FoldInstruction(tc.test_body, tc.id_to_fold, SPV_ENV_UNIVERSAL_1_1);
+  CheckForExpectedScalarConstant(
+      inst, tc.expected_result, [](const analysis::Constant* c) {
+        return c->AsScalarConstant()->GetU64BitValue();
+      });
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    TestCase, LongIntegerInstructionFoldingTest,
+    ::testing::Values(
+        // Test case 0: fold 1+4611686018427387904
+        InstructionFoldingCase<uint64_t>(
+            Header() + "%main = OpFunction %void None %void_func\n" +
+                "%main_lab = OpLabel\n" +
+                "%n = OpVariable %_ptr_int Function\n" +
+                "%load = OpLoad %int %n\n" +
+                "%2 = OpIAdd %long %long_1 %long_4611686018427387904\n" +
+                "OpReturn\n" + "OpFunctionEnd",
+            2, 1 + 4611686018427387904),
+        // Test case 1: fold 1-4611686018427387904
+        InstructionFoldingCase<uint64_t>(
+            Header() + "%main = OpFunction %void None %void_func\n" +
+                "%main_lab = OpLabel\n" +
+                "%n = OpVariable %_ptr_int Function\n" +
+                "%load = OpLoad %int %n\n" +
+                "%2 = OpISub %long %long_1 %long_4611686018427387904\n" +
+                "OpReturn\n" + "OpFunctionEnd",
+            2, 1 - 4611686018427387904),
+        // Test case 2: fold 2*4611686018427387904
+        InstructionFoldingCase<uint64_t>(
+            Header() + "%main = OpFunction %void None %void_func\n" +
+                "%main_lab = OpLabel\n" +
+                "%n = OpVariable %_ptr_int Function\n" +
+                "%load = OpLoad %int %n\n" +
+                "%2 = OpIMul %long %long_2 %long_4611686018427387904\n" +
+                "OpReturn\n" + "OpFunctionEnd",
+            2, 9223372036854775808ull),
+        // Test case 3: fold 4611686018427387904/2 (unsigned)
+        InstructionFoldingCase<uint64_t>(
+            Header() + "%main = OpFunction %void None %void_func\n" +
+                "%main_lab = OpLabel\n" +
+                "%n = OpVariable %_ptr_int Function\n" +
+                "%load = OpLoad %int %n\n" +
+                "%2 = OpUDiv %long %long_4611686018427387904 %long_2\n" +
```

Comment:
I updated the test cases. I guess we are not running the validator for these cases.

> Just want to make sure we don't create any cases where we could fold something invalid into something valid.

I don't think this is a problem unless we do something totally unexpected. We can leave it to the validator to identify problems. 

> Could you add a test for OpUDiv %ulong %long_a %long_b, since that should definitely not fold to an OpConstant %ulong?

This is invalid, and would be caught by the validator because the result type does not match the operands's type. 
I do not want to assert because I want to use generic code that covers other opcodes like OpIAdd, which do not have the same restriction. For OpIAdd, only the widths have to match.

Reply:
Isn't a more typical flow to run the optimizer before the validator though? At least in the case of DXC, if we had some bad codegen that produced `OpUDiv %ulong %long_a %long_b` (wrong but not a totally unexpected sort of bug), I don't think the optimizer should turn it into a valid `OpConstant`.

Maybe that is too DXC-focused a perspective, but if the assumption is that all code that goes into the optimizer should be validated beforehand, I don't think we currently do that.

Reply:
The interface for the optimizer says that it will run the validator itself on the original binary. If you use the interface to turn of the validator, then you are asserting that your binary is valid. See https://github.com/KhronosGroup/SPIRV-Tools/blob/80926d97aba8d1b1162448713ca4825d4c816226/include/spirv-tools/optimizer.hpp#L179-L221

Reply:
SG, thanks!